aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+
+    return (
+        (1 + HAIL_SERVICE_FEE)
+        * currency_conversion_rate
+        * get_usd_cost_for_resource(batch_resource, usage)
+    )
+
+
+def get_bq_schema_json() -> dict[str, any]:
+    """Get the schema for the table"""
+    pwd = Path(__file__).parent.parent.resolve()
+    schema_path = pwd / 'schema' / 'aggregate_schema.json'
+    with open(schema_path, 'r', encoding='utf-8') as f:
+        return json.load(f)
+
+
+def _format_bq_schema_json(schema: dict[str, Any]):
+    """
+    Take bq json schema, and convert it to bq.SchemaField objects"""
+    formatted_schema = []
+    for row in schema:
+        kwargs = {
+            'name': row['name'],
+            'field_type': row['type'],
+            'mode': row['mode'],
+        }
+
+        if 'fields' in row and row['fields']:
+            kwargs['fields'] = _format_bq_schema_json(row['fields'])
+        formatted_schema.append(bq.SchemaField(**kwargs))
+    return formatted_schema
+
+
+def get_formatted_bq_schema() -> list[bq.SchemaField]:
+    """
+    Get schema for bigquery billing table, as a list of bq.SchemaField objects
+    """
+    return _format_bq_schema_json(get_bq_schema_json())
+
+
+def parse_hail_time(time_str: str) -> datetime:
+    """
+    Parse hail datetime object
+
+    >>> parse_hail_time('2022-06-09T04:59:58Z').isoformat()
+    '2022-06-09T04:59:58'
+    """
+    if isinstance(time_str, datetime):
+        return time_str
+
+    return datetime.strptime(time_str, '%Y-%m-%dT%H:%M:%SZ')
+
+
+def to_bq_time(time: datetime):
+    """Convert datetime to transport datetime for bigquery"""
+    return time.strftime('%Y-%m-%d %H:%M:%S')
+
+
+def get_hail_token() -> str:
+    """
+    Get Hail token from local tokens file
+    """
+    if os.getenv('DEV') in ('1', 'true', 'yes'):
+        with open(os.path.expanduser('~/.hail/tokens.json'), encoding='utf-8') as f:
+            config = json.load(f)
+            return config['default']
+
+    assert GCP_PROJECT
+    return read_secret(GCP_PROJECT, 'aggregate-billing-hail-token')
+
+
+def get_credits(
+    entries: Iterable[dict[str, Any]],
+    topic: str,
+    project: dict,
+) -> list[dict[str, any]]:
+    """
+    Get a hail/seqr credit for each entry.
+    """
+
+    hail_credits = [{**e} for e in entries]
+    for entry in hail_credits:
+        entry['topic'] = topic
+        entry['id'] += '-credit'
+        entry['cost'] = -entry['cost']
+        entry['service'] = {
+            **entry['service'],
+            'description': entry['service']['description'] + ' Credit',
+        }
+        sku = {**entry['sku']}
+        sku['id'] += '-credit'
+        sku['description'] += '-credit'
+        entry['sku'] = sku
+        entry['project'] = project
+
+    return hail_credits
+
+
+async def get_batches(
+    token: str,
+    billing_project: str | None = None,
+    last_batch_id: Any | None = None,
+) -> dict[str, any]:
+    """
+    Get list of batches for a billing project with no filtering.
```

Comment:
What happens if `billing_project` is None?

Reply:
The behaviour of the endpoint is to return all the user is a part of - we don't do anything to the request though.

Reply:
I've clarified in a comment :)